src/core/control/Control.cpp

```diff
@@ -1087,6 +1093,9 @@ void Control::toolChanged() {
     }
 
     getCursor()->updateCursor();
+    if (win) {
+        win->getOpacityToolbox()->update();
+    }
```

Comment:
@bhennion,
Wow, I felt a bit concerned by putting so many `opacityToolbox->update()` through all over the place and It looked quite unmaintenable and messy.

I thought these GActions were only connected to menu and buttons and did not know you could get notified anywhere from the code when they are triggered: this perfectly fits to my case.

I'm really amazed how things get always cleaner and optimized with you. Thank you very much.

Right now, I have a little problem.

given the following code :
```c++
ActionRef gActionSelectTool = actionDB->getAction(Action::SELECT_TOOL);

g_signal_connect(gActionSelectTool.get(), "notify::state", G_CALLBACK(+[](GObject* a, GParamSpec*, OpacityToolbox* self) {
		xoj::util::GVariantSPtr state(g_action_get_state(G_ACTION(a)), xoj::util::adopt);
		auto toolType = getGVariantValue<ToolType>(state.get());
		g_log("OpacityToolbox::gActionSelectTool", G_LOG_LEVEL_DEBUG, "toolType_from_state=%s ; toolType_from_toolhandler=%s", toolTypeToString(toolType).c_str(), toolTypeToString(self->toolHandler->getToolType()).c_str());
	}),
	this);
``` 
When I click on a button to select a new tool, I get notified before the new tool is selected via ctrl->selectTool() in the callback. 
```c++
/** Tool menu **/
template <>
struct ActionProperties<Action::SELECT_TOOL> {
    using state_type = ToolType;
    using parameter_type = state_type;
    static state_type initialState(Control* ctrl) { return ctrl->getToolHandler()->getActiveTool()->getToolType(); }
    static void callback(GSimpleAction* ga, GVariant* p, Control* ctrl) {
    	g_log("ActionProperties::SELECT_TOOL callback()", G_LOG_LEVEL_DEBUG, "--> entered");
        g_simple_action_set_state(ga, p);
        ToolType tt = getGVariantValue<ToolType>(p);
        xoj_assert(tt < TOOL_END_ENTRY);
        if (requiresClearedSelection(tt)) {
            ctrl->clearSelection();
        }
        g_log("ActionProperties::SELECT_TOOL callback()", G_LOG_LEVEL_DEBUG, "--> ctrl->selectTool");
        ctrl->selectTool(tt);
    }
};
``` 
So at this time, toolHandler is not yet in the updated state that I need to update the opacityToolbox accordingly.
However, since ctrl->selectTool() is invoked, and thus, also Control::toolChanged() ...
```c++
void Control::toolChanged() {
    g_log("Control::toolChanged", G_LOG_LEVEL_DEBUG, "--> entered");
    ToolType type = toolHandler->getToolType();

    this->actionDB->setActionState(Action::SELECT_TOOL, type);
```
Do you have an idea why I don't get notified from `this->actionDB->setActionState(Action::SELECT_TOOL, type);`? Is there a way to be notiified after the tool handler has already set the new tool, rather than just before?

Reply:
Thank you for your further explanations.

Now I understand why by inlining `ctrl->selectTool(tt);` in the callback, the infered `g_simple_action_set_state(ga, p)` will not notify my handler. This is because the state did not change between the two calls of `g_simple_action_set_state()`.
I mistakenly thought that a simple call to `g_simple_action_set_state` would notify the handler, but it actually requires a change in the state for the notification to occur.